src/elisp/treemacs.el

```diff
@@ -240,6 +240,97 @@ An error message is displayed if the current buffer is not part of any project."
              (treemacs-goto-file-node path)
            (treemacs-add-project-to-workspace path name)))))))
 
+;;;###autoload
+(defun treemacs--bookmark-handler (record)
+  "Open Treemacs into a bookmark RECORD."
+  (treemacs-unless-let (path (bookmark-prop-get record 'treemacs-bookmark-path))
+      (user-error "Treemacs--bookmark-handler invoked for a non-Treemacs bookmark")
+    (unless (window-live-p (treemacs-get-local-window))
+      (treemacs))
+    (select-window (treemacs-get-local-window))
+    (treemacs-goto-node path)
+    ;; If the user has bookmarked a directory, they probably want to operate on
+    ;; its contents. Expand it, and select the first child.
+    (when (treemacs-is-node-collapsed? (treemacs-current-button))
+      (treemacs-TAB-action))
+    (let ((depth (treemacs-button-get (treemacs-current-button) :depth))
+          (next-button (next-button (treemacs-current-button))))
+      (when (and next-button (> (treemacs-button-get next-button :depth) depth))
+        (treemacs-next-line 1)))))
+
+(defun treemacs--format-bookmark-title (btn)
+  "Format the bookmark title for BTN with `treemacs-bookmark-title-template'."
+  (s-format
```

Comment:
Now that's a mouthful. You should add a few tests, so I'll have examples for what input and output is supposed to look like.